system/jlib/jthread.cpp

```diff
@@ -702,22 +702,47 @@ void CAsyncFor::For(unsigned num,unsigned maxatonce,bool abortFollowingException
                 return 0;
             }
         };
-        if (maxatonce==0)
+        if (maxatonce==0 || maxatonce > num)
             maxatonce = num;
-        for (i=0;(i<num)&&(i<maxatonce);i++)
-            ready.signal();
-        IArrayOf<Thread> started;
-        started.ensure(num);
-        for (i=0;i<num;i++) {
-            ready.wait();
-            if (abortFollowingException && e) break;
-            Owned<Thread> thread = new cdothread(this,shuffled?shuffler->lookup(i):i,ready,&errmutex,e);
-            thread->start();
-            started.append(*thread.getClear());
+        if (maxatonce < num)
+        {
+            for (i=0;(i<num)&&(i<maxatonce);i++)
+                ready.signal();
+            IArrayOf<Thread> started;
+            started.ensure(num);
+            for (i=0;i<num;i++) {
+                ready.wait();
+                if (abortFollowingException && e) break;
+                Owned<Thread> thread = new cdothread(this,shuffled?shuffler->lookup(i):i,ready,&errmutex,e);
+                thread->start();
+                started.append(*thread.getClear());
+            }
+            ForEachItemIn(idx, started)
+            {
+                started.item(idx).join();
+            }
         }
-        ForEachItemIn(idx, started)
+        else
         {
-            started.item(idx).join();
+            // Common case of execute all at once can be optimized a little
+            // Note that shuffle and abortFollowingException are meaningless when executing all at once
+            dbgassertex(!shuffled);
```

Comment:
not sure this is a good idea.  Some general purpose code e.g.

    afor.For(eps.ordinality(), 10, false, true);

will trigger the assert if the list is too short

Reply:
True. Unlikely, but I will remove the asserts